cfgrammar/src/lib/yacc/grammar.rs

```diff
@@ -1025,14 +1025,14 @@ where
 
 #[derive(Debug)]
 pub enum YaccGrammarError {
```

Comment:
This might be a dumb question (I don't remember this bit of the code well enough): could/should we fold `YaccParserError` and `GrammarValidationError` into one? Or, perhaps, should we return `Vec<YaccGrammarError>`?

Reply:
I'm not terribly familiar with `GrammarValidationError` either, 

It seems plausible to me, the line between `GrammarValidationErrors` and `YaccParserErrors` is blurry.
For example, `Duplicate`* could be seen as validation errors, since they do not cause us to fail to parse later stages.

But `Duplicate`* can be checked as we parse, and doing so is beneficial because it means we don't add state which is known to be invalid, e.g. to the `rules`. field.

So, `GrammarValidationError` is kind of a niche thing, for errors which cannot be checked at parse time.
I.e. `NoStartRule` can't be determined until *after* we've parsed all the rules.  

If we take the position (hypothetically) that `Duplicate`* should really be validation errors.
We find that we are potentially returning errors, when we *know* about other errors.

I believe that the point/benefit of `GrammarValidationError` might just be that they *don't* get added checked,
Unless we successfully parsed.  I.e. If we hit a syntax error parsing a rule, we don't want to *also* emit a `NoStartRule` error which we could do if we moved the `GrammarValidationError` checks to `parse`.

This doesn't necessarily mean that they shouldn't or couldn't be in the same `enum`, but being separate we can't make this mistake of trying to perform grammar validation in parse, and emitting cascading errors that only fail because of previous errors.

I believe that is how I read the situation, anyways.

Reply:
Sorry for perhaps excessive detail, there is a counter-argument that I failed to notice above:

We currently don't emit `NoStartRule` when we could, for instance given:
```
%avoid_insert a
%avoid_insert a
%%
```
only when we fix the error we then get 'No Start Rule'.

```
%%
Start-Rule:;
```
The rule will fail to parse because of the `-` in the rule, but it looks for
bar.y: Missing ':' at line 2 column 6

And currently *won't* emit `NoStartRule`, and I would say arguably it shouldn't.
Currently the types prevent us from emitting `NoStartRule` in the first case.
I'm not exactly sure the heuristics involved with trying to emit it in the first while avoiding it in the second.

This seems like a decent argument for combining them though, and perhaps doing the checking in the
`Ok(i) => if errors.len() { ... } else { check start rule(...); Ok(...) } branch`

Reply:
Would it make sense to try combining them "before" this PR (if that makes sense)?

Reply:
From the perspective of doing both changes I don't think it makes much difference the order we do the changes in, the overlap is minimal, we're going to have to make 2 sets of changes either order we go.  

Unless what we're worried about is the affect on the enum size. like if for a currently unforeseen reason the second one we encounter an issue that causes us not to want to combine these.  And we don't want to end up with this enum layout it may not make sense to introduce it in the first place, and if that fails, we can think about what other alternatives we might have.

Reply:
> Unless what we're worried about is the affect on the enum size. like if for a currently unforeseen reason the second one we encounter an issue that causes us not to want to combine these.

Yes, that's what I'm (mildly) concerned about.

Reply:
Alright, I'll go take a stab at that first. I'll probably leave this open for now in hopes that the conflicts will be minimal enough that it's still salvageable :pray:

Reply:
Works for me!